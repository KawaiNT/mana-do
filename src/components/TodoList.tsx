import React, { useState } from 'react';
import { isTodoCompleted } from '../utils';
import { TodoStatus } from '../models/todo';
import { Todo, TodoListInterface } from '../models/todo';
import TodoEditForm from './TodoEditForm';

const TodoList = (props: TodoListInterface) => {
  const emptyTodo = { id: '', user_id: '', content: '', created_date: ''}
  const [todoEdit, setTodoEdit] = useState(emptyTodo);

  const showTodos = props.todos && props.todos.filter((todo) => {
    switch (props.showing) {
      case TodoStatus.ACTIVE:
        return todo.status === TodoStatus.ACTIVE;
      case TodoStatus.COMPLETED:
        return todo.status === TodoStatus.COMPLETED;
      default:
        return true;
    }
  });

  const onEdit = (todo: Todo) => {
    setTodoEdit(todo)
  }

  const onUpdateTodo = (todo: Todo) => {
    props.onUpdateTodoContent(todo.id, todo.content)
    setTodoEdit(emptyTodo)
  }

  const onCancelUpdate = () => {
    setTodoEdit(emptyTodo)
  }

  return (
    <div className="ToDo__list" data-testid="todo-list">
      { props.showing === 'ALL' && (<p>All to-do list: <strong>{ showTodos?.length }</strong></p>)}
      { props.showing === 'ACTIVE' && (<p>Active to-do list: <strong>{ showTodos?.length }</strong></p>)}
      { props.showing === 'COMPLETED' && (<p>Completed to-do list: <strong>{ showTodos?.length }</strong></p>)}
      { showTodos && showTodos.map((todo, index) => (
          <div key={todo.id}>
            {todoEdit.id === todo.id ? (
              <TodoEditForm
                todo={todo}
                onUpdateTodo={onUpdateTodo}
                onCancelUpdate={onCancelUpdate}
              />
            ) : (
              <div key={index} className="ToDo__item">
                <input
                  type="checkbox"
                  checked={isTodoCompleted(todo)}
                  onChange={(e) => props.onUpdateTodoStatus(e, todo.id)}
                />
                <span onClick={() => onEdit(todo)}>{todo.content}</span>
                <button
                  className="Todo__delete"
                  onClick={() => props.onDelete(todo.id)}
                >
                  X
                </button>
              </div>
            )}
          </div>
        ))
      }
    </div>
  )
}

export default TodoList