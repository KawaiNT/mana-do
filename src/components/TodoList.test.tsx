import React from 'react';
import { render } from '@testing-library/react';
import TodoList from './TodoList'

describe('TodoList component test', () => {
  test('TodoList render', () => {
    const onDelete = jest.fn()
    const onUpdateTodoStatus = jest.fn()
    const onUpdateTodoContent = jest.fn()
    const { asFragment } = render(<TodoList todos={[]} showing='ALL' onDelete={onDelete} onUpdateTodoContent={onUpdateTodoContent} onUpdateTodoStatus={onUpdateTodoStatus} />)

    expect(asFragment(<TodoList todos={[]} showing='ALL' onDelete={onDelete} onUpdateTodoContent={onUpdateTodoContent} onUpdateTodoStatus={onUpdateTodoStatus} />)).toMatchSnapshot()
  })

  test('TodoList should be in the document', () => {
    const onDelete = jest.fn()
    const onUpdateTodoStatus = jest.fn()
    const onUpdateTodoContent = jest.fn()
    const { getByTestId } = render(<TodoList todos={[]} showing='ALL' onDelete={onDelete} onUpdateTodoContent={onUpdateTodoContent} onUpdateTodoStatus={onUpdateTodoStatus} />)
    const todoList = getByTestId('todo-list')

    expect(todoList).toBeInTheDocument()
  })
})