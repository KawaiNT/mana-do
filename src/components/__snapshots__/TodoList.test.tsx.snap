// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component test TodoList render 1`] = `
<DocumentFragment>
  <div
    class="ToDo__list"
    data-testid="todo-list"
  >
    <p>
      All to-do list: 
      <strong>
        0
      </strong>
    </p>
  </div>
</DocumentFragment>
`;
